plutus-benchmark/bench-validation/Main.hs

```diff
@@ -0,0 +1,142 @@
+
+import           Codec.Serialise
+import           Control.Monad
+import           Control.Monad.Trans.Except                                 (runExceptT)
+import qualified Data.ByteString.Lazy                                       as BSL
+import           Data.Functor                                               ((<&>))
+import qualified Data.Map                                                   as Map
+import           System.FilePath
+import           Text.Printf                                                (printf)
+
+import qualified Language.PlutusCore                                        as PLC
+import           Language.PlutusCore.CBOR
+import           Language.PlutusCore.Constant                               (DynamicBuiltinNameMeanings (..))
+import           Language.PlutusCore.Constant.Dynamic
+import           Language.PlutusCore.Evaluation.Machine.ExBudgetingDefaults
+import qualified Language.PlutusCore.Universe                               as PLC
+import qualified Language.UntypedPlutusCore                                 as UPLC
+import qualified Language.UntypedPlutusCore.DeBruijn                        as UPLC
+import qualified Language.UntypedPlutusCore.Evaluation.Machine.Cek          as UPLC
+
+type Term a    = UPLC.Term PLC.Name PLC.DefaultUni a
+type Program a = UPLC.Program PLC.Name PLC.DefaultUni a
+
+-- Generate an HTML report.  If run via stack/cabal this will be written to the
+-- `plutus-benchmark` directory.
+config :: Config
+config = defaultConfig
+  { reportFile = Just "report.html"
+  }
+
+fromDeBruijn ::  UPLC.Program UPLC.DeBruijn PLC.DefaultUni a ->  IO (Program a)
+fromDeBruijn prog = do
+    let namedProgram = UPLC.programMapNames (\(UPLC.DeBruijn ix) -> UPLC.NamedDeBruijn "v" ix) prog
+    case PLC.runQuote $ runExceptT $ UPLC.unDeBruijnProgram namedProgram of
+      Left e  -> error $ show e
+      Right p -> return p
+
+loadPlc :: Serialise a => FilePath -> IO (Program a)
+loadPlc file = do
+  BSL.readFile file <&> deserialiseOrFail >>= mapM fromDeBruijn >>= \case
+               Left (DeserialiseFailure offset msg) ->
+                   error ("Deserialisation failure at offset " ++ Prelude.show offset ++ ": " ++ msg)
+               Right r -> return r
+
+benchCek :: Term () -> Benchmarkable
+benchCek program = nf (UPLC.unsafeEvaluateCek getStringBuiltinMeanings defaultCostModel) program
+
+cborSuffix :: String
+cborSuffix = ".cbor"
+
+dataDir :: String
+dataDir = "bench-validation" </> "data"
+
+{- Construct an applied validator.  We assume that relevant validators, datum
+   scripts, redeemers and contexts are stored in CBOR format under `<progName>`
+   in the `data` directory.  These should have names like "Redeemer01.cbor",
+   "Context03.cbor", and so on. This function returnes a Criterion environment
+   to be fed to the relevant benchmark, to keep the IO overhead out of the
+   benchmark itself. -}
+getAppliedScript :: String -> Int -> Int -> Int -> Int -> IO (Term ())
```

Comment:
I am sure there is a good reason for it, but why are we using files to store everything (rather than generate them on the fly)?

Reply:
I'd somehow have to get the blockchain simulator to run the tests and save the scripts just before it's about to validate them (and in suitably named files).  I think that would require quite big changes to the ledger code, so this seemed like the easiest way to get started (and it was reasonably quick).  It also means that we've got concrete files that you and Simon can experiment with.

Reply:
We might want to find a better way of doing this later, but this at least gets us started.  

I guess there's also a question of what we're actually trying to benchmark.  If we generate the validators every time we run the benchmarks then (for example) changes to the optimiser might change the speed of the benchmarks, but that would make it difficult to compare different versions of the machine if they were on different branches and the rest of the infrastructure was also changing.  Maybe we need some fixed benchmarks for the evaluator and some that are always up to date for testing other aspects of the system.

Reply:
If there is a good way to generate the golden files (and there is one), I think this is a good first solution.

Reply:
> If there is a good way to generate the golden files (and there is one), I think this is a good first solution.

I'm not quite sure what you mean there, @raduom (in particular, which golden files you're referring to).  Could you say a little more?

Reply:
When I initially saw the PR there were a couple of questions that popped up for me. The first one is: What if the compiler gets changed (some optimisation, for example) and outputs different code? The one that made the changes has to be aware that there are some files that need to be modified, and has to update them.

Reply:
@raduom I agree that it looks like maintaining the scripts is going to be hard. I am currently making some changes to the emulator that should make it easier to extract scripts at the point where they are evaluated. It's not something that @kwxm 's PR should be held up on, but I think we will be able to move away from having scripts stored in files pretty soon. 

So I am inclined to keep it as is in the current PR and then update it later on when we have a better way of extracting the scripts.

Reply:
> @raduom I agree that it looks like maintaining the scripts is going to be hard. I am currently making some changes to the emulator that should make it easier to extract scripts at the point where they are evaluated. 

That would be very helpful.  I think we'd probably still need to do some post-processing to eliminate duplicates: for example, in the multisig state machine test there are 28 validations, each involving a validator, datum, redeemer, and context.  However, only 10 different validations occur, involving one unique validator, 5 datum scripts, 5 redeemers and 10 contexts.  I had to extract 112 scripts and then compare them with a program and select the unique scripts manually, which was slightly annoying.  I suppose we can write a program to do all of the automatically though.  I'm still not entirely sure how the scripts  I have correspond to the events occurring during the progress of the contracts.